### ISSUE TITLE
[GF-JDK9] Exception when starting GF with JDK9

### ISSUE THREAD

**Issue by glassfishrobot** [eclipse-ee4j/glassfish#21588]
Before starting GF please set the following environment variables.
export JAVA_HOME = <JDK9Home>
export AS_JAVA= $JAVA_HOME
Now when GF is started we get the following exception in CLI
Exception in thread "main" java.lang.NullPointerException
at com.sun.enterprise.module.common_impl.AbstractModulesRegistryImpl.initializeServiceLocator(AbstractModulesRegistryImpl.java:152)
at com.sun.enterprise.module.common_impl.AbstractModulesRegistryImpl.newServiceLocator(AbstractModulesRegistryImpl.java:144)
at com.sun.enterprise.module.common_impl.AbstractModulesRegistryImpl.createServiceLocator(AbstractModulesRegistryImpl.java:218)
at com.sun.enterprise.module.common_impl.AbstractModulesRegistryImpl.createServiceLocator(AbstractModulesRegistryImpl.java:224)
at com.sun.enterprise.module.single.StaticModulesRegistry.createServiceLocator(StaticModulesRegistry.java:88)
at com.sun.enterprise.admin.cli.CLIContainer.getServiceLocator(CLIContainer.java:216)
at com.sun.enterprise.admin.cli.CLIContainer.getLocalCommand(CLIContainer.java:254)
at com.sun.enterprise.admin.cli.CLICommand.getCommand(CLICommand.java:231)
at com.sun.enterprise.admin.cli.AdminMain.executeCommand(AdminMain.java:369)
at com.sun.enterprise.admin.cli.AdminMain.doMain(AdminMain.java:304)
at org.glassfish.admin.cli.AsadminMain.main(AsadminMain.java:56)
-----------------------------------------------------------------------------------------------------------------------------------------------------------
Using some debug log we get the Root cause of the exception
java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: javax/annotation/PostConstruct
at java.util.concurrent.FutureTask.report(java.base@9-ea/FutureTask.java:122)
at java.util.concurrent.FutureTask.get(java.base@9-ea/FutureTask.java:191)
at org.glassfish.hk2.utilities.cache.LRUHybridCache$OriginThreadAwareFuture.get(LRUHybridCache.java:164)
at org.glassfish.hk2.utilities.cache.LRUHybridCache.compute(LRUHybridCache.java:303)
at org.glassfish.hk2.utilities.reflection.internal.ClassReflectionHelperImpl.findPostConstruct(ClassReflectionHelperImpl.java:119)
at org.jvnet.hk2.internal.Utilities.findPostConstruct(Utilities.java:1370)
at org.jvnet.hk2.internal.DefaultClassAnalyzer.getPostConstructMethod(DefaultClassAnalyzer.java:130)
at org.jvnet.hk2.internal.Utilities.getPostConstruct(Utilities.java:279)
at org.jvnet.hk2.internal.ClazzCreator.initialize(ClazzCreator.java:168)
at org.jvnet.hk2.internal.ClazzCreator.initialize(ClazzCreator.java:179)
at org.jvnet.hk2.internal.SystemDescriptor.internalReify(SystemDescriptor.java:723)
at org.jvnet.hk2.internal.SystemDescriptor.reify(SystemDescriptor.java:678)
at org.jvnet.hk2.internal.ServiceLocatorImpl.reifyDescriptor(ServiceLocatorImpl.java:459)
at org.jvnet.hk2.internal.ServiceLocatorImpl.narrow(ServiceLocatorImpl.java:2246)
at org.jvnet.hk2.internal.ServiceLocatorImpl.igdCacheCompute(ServiceLocatorImpl.java:1162)
at org.jvnet.hk2.internal.ServiceLocatorImpl.access$400(ServiceLocatorImpl.java:123)
at org.jvnet.hk2.internal.ServiceLocatorImpl$8.compute(ServiceLocatorImpl.java:1156)
at org.jvnet.hk2.internal.ServiceLocatorImpl$8.compute(ServiceLocatorImpl.java:1153)
at org.glassfish.hk2.utilities.cache.internal.WeakCARCacheImpl.compute(WeakCARCacheImpl.java:116)
at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetDescriptor(ServiceLocatorImpl.java:1237)
at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:768)
at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:735)
at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:705)
at com.sun.enterprise.module.common_impl.AbstractModulesRegistryImpl.initializeServiceLocator(AbstractModulesRegistryImpl.java:149)
at com.sun.enterprise.module.common_impl.AbstractModulesRegistryImpl.newServiceLocator(AbstractModulesRegistryImpl.java:144)
at com.sun.enterprise.module.common_impl.AbstractModulesRegistryImpl.createServiceLocator(AbstractModulesRegistryImpl.java:218)
at com.sun.enterprise.module.common_impl.AbstractModulesRegistryImpl.createServiceLocator(AbstractModulesRegistryImpl.java:224)
at com.sun.enterprise.module.single.StaticModulesRegistry.createServiceLocator(StaticModulesRegistry.java:88)
at com.sun.enterprise.admin.cli.CLIContainer.getServiceLocator(CLIContainer.java:216)
at com.sun.enterprise.admin.cli.CLIContainer.getLocalCommand(CLIContainer.java:254)
at com.sun.enterprise.admin.cli.CLICommand.getCommand(CLICommand.java:231)
at com.sun.enterprise.admin.cli.AdminMain.executeCommand(AdminMain.java:369)
at com.sun.enterprise.admin.cli.AdminMain.doMain(AdminMain.java:304)
at org.glassfish.admin.cli.AsadminMain.main(AsadminMain.java:56)
Caused by: java.lang.NoClassDefFoundError: javax/annotation/PostConstruct
at org.glassfish.hk2.utilities.reflection.internal.ClassReflectionHelperUtilities.isPostConstruct(ClassReflectionHelperUtilities.java:200)
at org.glassfish.hk2.utilities.reflection.internal.ClassReflectionHelperImpl.getPostConstructMethod(ClassReflectionHelperImpl.java:186)
at org.glassfish.hk2.utilities.reflection.internal.ClassReflectionHelperImpl.access$200(ClassReflectionHelperImpl.java:56)
at org.glassfish.hk2.utilities.reflection.internal.ClassReflectionHelperImpl$1.compute(ClassReflectionHelperImpl.java:64)
at org.glassfish.hk2.utilities.reflection.internal.ClassReflectionHelperImpl$1.compute(ClassReflectionHelperImpl.java:60)
at org.glassfish.hk2.utilities.cache.LRUHybridCache$OriginThreadAwareFuture$1.call(LRUHybridCache.java:115)
at org.glassfish.hk2.utilities.cache.LRUHybridCache$OriginThreadAwareFuture$1.call(LRUHybridCache.java:111)
at java.util.concurrent.FutureTask.run(java.base@9-ea/FutureTask.java:264)
at org.glassfish.hk2.utilities.cache.LRUHybridCache$OriginThreadAwareFuture.run(LRUHybridCache.java:173)
at org.glassfish.hk2.utilities.cache.LRUHybridCache.compute(LRUHybridCache.java:292)
... 30 more
Caused by: java.lang.ClassNotFoundException: javax.annotation.PostConstruct
at jdk.internal.loader.BuiltinClassLoader.loadClass(java.base@9-ea/BuiltinClassLoader.java:366)
at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(java.base@9-ea/ClassLoaders.java:185)
at java.lang.ClassLoader.loadClass(java.base@9-ea/ClassLoader.java:419)
... 40 more
-----------------------------------------------------------------------------------------------------------------------------------------------------------
#### Affected Versions
[4.1]

**Comment by keilw** [comment]
Sounds like a duplicate of #21588

**Comment by bshannon** [comment]
This *is* #21588,

Did you mean this is a duplicate of some other bug?
Or some other bug is a duplicate of this?

And to be clear, GlassFish 5 is not supported with JDK 9+ and is not expected to work.

**Comment by keilw** [comment]
No idea which the other one was, there are far too many open issues, but there was another one with "Java 10 and 9 throws NPE" or similar. 

How long is the incompatibility with Java 9 or 12/13 likely to last? If there ever was a Glassfish 5.x (e.g. 5.2) for Jakarta EE.next would it also suffer from this issue?

There is certainly a lot of work to be done in the documentation, READMEs and other documents all still mention the old Java EE sites now archived. 
And the mentioned problem even lasts when `JAVA_HOME` is set to Java 8, at least on Windows. Looks like Glassfish ignores `JAVA_HOME ` and only relies on its own `AS_JAVA ` variable. That should be clearly documented somewhere in the README IMO.

**Comment by bshannon** [comment]
GlassFish won't support JDK 9+ until someone starts a project to fix it.
In particular, I don't expect it to be fixed for GlassFish 5.2 / Jakarta EE 8.

I don't expect a fix to be *required* until GlassFish needs to support Jakarta EE 9.

If you want to see updates to the READMEs, please submit PRs.

And of course if you want to do the work to make GlassFish work with JDK 9+, we'd love to see PRs for that too!

**Comment by keilw** [comment]
I guess I'll leave JDK 9, 10 or 11 support to those who have their hands in OpenJDK, but improving the usability and experience by pointing out e.g. the necessary tweaks in the README, I guess I'll try that for now.

**Comment by keilw** [comment]
This one's a duplicate: #22460
